### ISSUE TITLE
BinaryWriter is actually faster in long run - comparing BinaryWriter, ArrayBuffer+DataView and DynamicBuffer perfomance

### ISSUE THREAD

**Issue by Luka967** [Barbosik/MultiOgar#203]
So I was bored after making a small representation of buffers called ByteBuffer which ended up extremely slow in perfomance. I came along with an idea to compare all buffers we've made to now, along with the old companions - Buffer, that slow guy ByteBuffer, ArrayBuffer + DataView combo and DynamicBuffer.

I made a single modification on DynamicBuffer which removes `flush` function, because it was actually useless as `moveTemp` function reads only the necessary byte amount. With the flush function time will exceed 100ms xD

The test was to write an unisigned 32-bit integer with value 2147483647, 100 000 times. The specs ain't awesome, it was done on a c9 virtual machine (512 MB ram, ~1GHz processor or even slower), but that's another topic.

At my suprise, DynamicBuffer was actually faster than BinaryWriter - sometimes even 10 milliseconds faster!
![ss1](https://cloud.githubusercontent.com/assets/15956695/16899636/3515d820-4c0b-11e6-9d30-978717993f78.png)

Also, BinaryWriter uses the most RAM - because it uses functions instead of already calculated bytes.  Buffer predates ArrayBuffer for 4-5 years, cannot write 64-bit values, and has a security flaw.

Now someone is thinking "That's all cool, but those are 100 thousand operations, not the usual 20-200!". Operation number is just an exponent for the effectiveness of a class. More operations equals more precision on who's better.

This was just a test out of my boredom and I just wanted to share the results.

EDIT: I got even more bored and set up testing for buffer creation. Results again unexpected:
![ss2](https://cloud.githubusercontent.com/assets/15956695/16899804/3862cfa0-4c11-11e6-8d28-dddd81aedd7e.png)
DynamicBuffer is 2-3x faster when creating small buffers.

tl;dr From 5 same tests, approximate speeds are:
Buffer was the best in 3ms
ArrayBuffer+DataView was 2nd in 10ms
DynamicBuffer was 3rd in 42ms
BinaryWriter was 4th in 47ms
ByteBuffer was the slowest in 300ms

**Comment by Barbosik** [comment]
yes, BinaryWritter is a 2x times slower than Buffer. Because it allocate list to store all write operations.
Without it, code will be unreadable and it will be too hard to catch the bug.
But it allows to simplify code very significantly. Currently I didn't find more fast way for dynamic allocation.

**Comment by Luka967** [comment]
You can simply store bytes into an array, use a temporary Buffer to calculate the bytes, then copy required byte amount to the array. Just as used in DynamicBuffer. Will be even faster as Buffer needs less time to write bytes on physical memory.

**Comment by Barbosik** [comment]
@Luka967: _Buffer predates ArrayBuffer for 4-5 years, cannot write 64-bit values, and has a security flaw._

1) Buffer is much more efficient than ArrayBuffer. It has much better performance and your test approve it.

2) java script doens't support 64-bit values (there is jsavascript issue with 64-bit values, and it not related to Buffer or ArrayBuffer)

3) What security flaw you're talking about? I never heard about it

**Comment by Barbosik** [comment]
@Luka967: _You can simply store bytes into an array_

As I remember, I tried to store bytes instead of lambda. But it shows worse performance. So, I used more fast version with lambdas.

**Comment by Barbosik** [comment]
And I think that your test is incorrect. Probably it performs test with no take into account that JIT compiller is smart enough and can just ommit some parts of code which result is not used.

So, I just made a small test. And here is result:
```
G:\GIT\TestBuffer>node app.js
[INFO ] ===PASS 1===
[INFO ] DynamicBuffer: 7586.255 [ms] (check=16000)
[INFO ] BinaryWriter:  668.871 [ms] (check=16000)
[INFO ] ===PASS 2===
[INFO ] DynamicBuffer: 7982.670 [ms] (check=16000)
[INFO ] BinaryWriter:  657.101 [ms] (check=16000)
[INFO ] ===PASS 3===
[INFO ] DynamicBuffer: 7932.713 [ms] (check=16000)
[INFO ] BinaryWriter:  691.305 [ms] (check=16000)
[INFO ] ===PASS 4===
[INFO ] DynamicBuffer: 7929.353 [ms] (check=16000)
[INFO ] BinaryWriter:  659.713 [ms] (check=16000)
[INFO ] ===PASS 5===
[INFO ] DynamicBuffer: 7807.980 [ms] (check=16000)
[INFO ] BinaryWriter:  677.840 [ms] (check=16000)
[DEBUG] process.exit(0)
``` 

As you can see, DynamicBuffer is 10 times slower than BinaryWriter.
BinaryWriter takes **just 0,7 second** to serialize all data.
DynamicBuffer takes **7,5 seconds** to serialize all data. 

I'm used latest DynamicBuffer source from ogar project (https://github.com/OgarProject/Ogar/commit/0bbddbe5a55f9bb7afbf254e0aacfbcb0a925da0).

Here is a test source code:
```
function getTime(t) {
    return t[0] * 1000 + t[1] / 1000000;
}

for (var pass = 1; pass <= 5; pass++) {
    Logger.info("===PASS " + pass + "===");
    
    /* DynamicBuffer test */
    var check = 0;
    var t1 = process.hrtime();
    for (var i = 0; i < 1000; i++) {
        var writer = new DynamicBuffer(true);
        
        // cell update simulation
        writer.setUint8(16);
        for (var j = 0; j < 1000; j++) {
            writer.setUint32(0);    // ID
            writer.setInt32(0);     // X
            writer.setInt32(0);     // Y
            writer.setUint16(0);    // Size
            writer.setUint8(0);     // Flags
            writer.setUint8(0);     // r
            writer.setUint8(0);     // g
            writer.setUint8(0);     // b
        }
        var result = new Uint8Array(writer.build());
        // we create Uint8Array (because it anyway will be performed before pass the data to WebSocket) 
        // So, we take it into account
        result = new Uint8Array(result);
        // utilize result to prevent JIT optimization
        check += result[0];
    }
    var t2 = process.hrtime(t1);
    Logger.info("DynamicBuffer: " + getTime(t2).toFixed(3) + " [ms] (check=" + check + ")");
    
    
    
    /* BinaryWriter test */
    var check = 0;
    var t1 = process.hrtime();
    for (var i = 0; i < 1000; i++) {
        var writer = new BinaryWriter();
        
        // cell update simulation
        writer.writeUInt8(16);
        for (var j = 0; j < 1000; j++) {
            writer.writeUInt32(0);  // ID
            writer.writeInt32(0);   // X
            writer.writeInt32(0);   // Y
            writer.writeUInt16(0);  // Size
            writer.writeUInt8(0);   // Flags
            writer.writeUInt8(0);   // r
            writer.writeUInt8(0);   // g
            writer.writeUInt8(0);   // b
        }
        var result = writer.toBuffer();
        // utilize result to prevent JIT optimization
        check += result[0];
    }
    var t2 = process.hrtime(t1);
    Logger.info("BinaryWriter:  " + getTime(t2).toFixed(3) + " [ms] (check=" + check + ")");
}
```

**Comment by Luka967** [comment]
@Barbosik 
1. True. But it's older than ArrayBuffer, and is probably more susceptible to problems.
3. Right that. Unsafe allocation won't reset all allocated bytes to zero, so you can practically read the memory. However node.js isn't a browser and reading own memory is probably a ok.
4. Number in javascript is a single-precision floating-point number (32-bit float). Lambdas/functions are probably stored as strings, or something else which uses more memory.
5.
> I made a single modification on DynamicBuffer which removes flush function, because it was actually useless as moveTemp function reads only the necessary byte amount. With the flush function time will exceed 100ms xD
Pretty sure you've forgot this.

**Comment by Barbosik** [comment]
@Luka967

_>>1. True. But it's older than ArrayBuffer, and is probably more susceptible to problems._

Buffer especially optimized for binary serialization for socket operations. So, if you want best performance, you should use Buffer. Other solutions is too slow for such operations.

_>>3. Right that. Unsafe allocation won't reset all allocated bytes to zero, so you can practically read the memory. However node.js isn't a browser and reading own memory is probably a ok._

BinaryWriter using safe bufer allocation. If you want fast unsafe allocation you need to allocate buffer in a little different way than just new Buffer(). So, it's for your choice what is better - allocation of buffer filled with zero or just allocation of buffer with some garbage.

_>>4. Number in javascript is a single-precision floating-point number (32-bit float). Lambdas/functions are probably stored as strings, or something else which uses more memory._

I don't know how node.js JIT compiller store functions. But usually JIT compillers using native pointer to store functions. Native pointer == platform word - it's the most fastest chunk of data. Also read/write of platform word is atomic operation and can be executed with no thread synchronization. So, there is no more fast way to store any kind of data. It's the most fastest type of data.

The problem with BinaryWriter is that it allocates _writers array. Allocation takes some time. It's not so significant. But it's the reason why BinaryWriter is 2x times slower than direct Buffer manipulation. 

_>>5.Pretty sure you've forgot this._

Yes, I didn't take it into account, because this pull request is still not applied to the code.
Just tested DynamicBuffer with this optimization, here is my results:
```
[INFO ] ===PASS 1===
[INFO ] DynamicBuffer: 2331.093 [ms] (check=16000)
[INFO ] BinaryWriter:  666.300 [ms] (check=16000)
[INFO ] ===PASS 2===
[INFO ] DynamicBuffer: 2396.094 [ms] (check=16000)
[INFO ] BinaryWriter:  642.268 [ms] (check=16000)
[INFO ] ===PASS 3===
[INFO ] DynamicBuffer: 2301.796 [ms] (check=16000)
[INFO ] BinaryWriter:  643.900 [ms] (check=16000)
[INFO ] ===PASS 4===
[INFO ] DynamicBuffer: 2328.291 [ms] (check=16000)
[INFO ] BinaryWriter:  645.757 [ms] (check=16000)
[INFO ] ===PASS 5===
[INFO ] DynamicBuffer: 2310.027 [ms] (check=16000)
[INFO ] BinaryWriter:  644.521 [ms] (check=16000)
[DEBUG] process.exit(0)
```

Your optimization really helps to improve performace up to 2x times.
But DynamicBuffer still much slower than BinaryWriter.
As you can see, BinaryWriter takes just 640 ms, but DynamicBuffer takes 2300 ms...

**Comment by Luka967** [comment]
Huh, actually true. Tested your code. I kept getting results being DynamicBuffer faster than BinaryWriter - no idea why. I'll put up more tests and also improve this ByteBuffer that I started, maybe it ends up well.

**Comment by Barbosik** [comment]
@Luka967: may be you're run it under debugger? If you running code under debugger it works a little different. You need to run performance tests without debugger and profiling options. Run it just with clean `node app.js`

**Comment by Luka967** [comment]
Nope, clean `node server.js` on terminal (the file name). I went by using simple Date subraction but still unsure what went different.

**Comment by Barbosik** [comment]
Using Date for performance measurement is not good way. Because Date has low resolution and may return cached value, so it will leads to incorrect measurement.

Use process.hrtime to get the time with high resolution. It works more fast and returns more precise time. It returns time in nanoseconds. The value in nanoseconds is too large to fit into 32 bit value, so it returns object with high and low part of the value.

**Comment by Luka967** [comment]
Thing is, Date is still good enough. I replaced all `process.hrtime` calls into Date calculations and they still turned out correct.
Apparently, DynamicBuffer needs a lot more time to create itself. Only that part is 0.2 ms. Creating 1000 of them is a 4 ms disadvantage. It's probably because of the temporary buffers used for calculation.

I was also digging into the `ws` module and found out that `ws` module accepts ArrayBuffer and any typed array and doesn't convert it into an Uint8Array. Just consider this code:
```js
  if (typeof options.binary === 'undefined') {
    options.binary = (data instanceof ArrayBuffer || data instanceof Buffer ||
      data instanceof Uint8Array ||
      data instanceof Uint16Array ||
      data instanceof Uint32Array ||
      data instanceof Int8Array ||
      data instanceof Int16Array ||
      data instanceof Int32Array ||
      data instanceof Float32Array ||
      data instanceof Float64Array);
  }
```
Copied from the `WebSocket.send` function. If you want to find it, /node_modules/ws/lib/WebSocket.js line 233.

**Comment by Barbosik** [comment]
@Luka967: _I was also digging into the ws module and found out that ws module accepts ArrayBuffer and any typed array and doesn't convert it into an Uint8Array._

You need to read and use result in the test. So, there is need to create Uint8Array to access result. If you will not use result, then node.js JIT compiller may just omit some parts of code because their results is not used. And you will measure just empty for loop :)

But anyway this doesn't affect test results. I tested it with new Uint8Array and without it - the result is almost the same

**Comment by Luka967** [comment]
When I removed conversion to Uint8Array, I got a 400ms difference xD

**Comment by Barbosik** [comment]
As I mention before - it's because JIT optimization. 
If you remove Uint8Array, you will not be able to access it's element to sum it into check value.
JIT compiller will found that this array is not used at all and just will skip code which creates it.

**Comment by Barbosik** [comment]
@Luka967: 
_>>Thing is, Date is still good enough. I replaced all process.hrtime calls into Date calculations and they still turned out correct._

Date is very slow, I found it with node.js profiler. It's much slower than process.hrtime. 
Also it has very low resolution. You cannot measure intervals with precision 1 ms by using Date, because you will get error +- 20 ms.
So, I don't recommend to use it for time interval measurements. It's useful if you need to get current time. But if you need to measure time interval, then use process.hrtime. It's more fast and allows to measure time more precisely (in nanoseconds).

**Comment by Barbosik** [comment]
More test results from different systems and node versions:

**Windows 7 x64, node.js v4.4.6:**
```
[INFO ] ===PASS 1===
[INFO ] DynamicBuffer: 2319.889 [ms] (check=16000)
[INFO ] BinaryWriter:  656.825 [ms] (check=16000)
[INFO ] ===PASS 2===
[INFO ] DynamicBuffer: 2353.371 [ms] (check=16000)
[INFO ] BinaryWriter:  625.612 [ms] (check=16000)
[INFO ] ===PASS 3===
[INFO ] DynamicBuffer: 2346.758 [ms] (check=16000)
[INFO ] BinaryWriter:  630.342 [ms] (check=16000)
[INFO ] ===PASS 4===
[INFO ] DynamicBuffer: 2350.783 [ms] (check=16000)
[INFO ] BinaryWriter:  630.305 [ms] (check=16000)
[INFO ] ===PASS 5===
[INFO ] DynamicBuffer: 2349.300 [ms] (check=16000)
[INFO ] BinaryWriter:  628.250 [ms] (check=16000)
[DEBUG] process.exit(0)
```

**Ubuntu x32, node.js v4.2.6:**
```
[INFO ] ===PASS 1===
[INFO ] DynamicBuffer: 2414.474 [ms] (check=16000)
[INFO ] BinaryWriter:  629.780 [ms] (check=16000)
[INFO ] ===PASS 2===
[INFO ] DynamicBuffer: 2345.037 [ms] (check=16000)
[INFO ] BinaryWriter:  621.148 [ms] (check=16000)
[INFO ] ===PASS 3===
[INFO ] DynamicBuffer: 2338.556 [ms] (check=16000)
[INFO ] BinaryWriter:  609.316 [ms] (check=16000)
[INFO ] ===PASS 4===
[INFO ] DynamicBuffer: 2325.968 [ms] (check=16000)
[INFO ] BinaryWriter:  608.318 [ms] (check=16000)
[INFO ] ===PASS 5===
[INFO ] DynamicBuffer: 2350.982 [ms] (check=16000)
[INFO ] BinaryWriter:  621.095 [ms] (check=16000)
```

**Comment by Barbosik** [comment]
And here is profiling data for test.
Both BinaryWriter and DynamicBuffer were executed, so we can compare speed of most slow function for both implementations:
```
 [JavaScript]:
   ticks  total  nonlib   name
   1169    7.3%    7.4%  LazyCompile: *setUint8 native typedarray.js:1843:28
    847    5.3%    5.4%  Stub: CEntryStub
    826    5.2%    5.2%  LazyCompile: *getUint8 native typedarray.js:1834:28
    814    5.1%    5.1%  LazyCompile: *DynamicBuffer.build G:\GIT\TestBuffer\DynamicBuffer.js:99:42
    678    4.2%    4.3%  Builtin: ArgumentsAdaptorTrampoline
    467    2.9%    3.0%  LazyCompile: *DynamicBuffer.setInt32 G:\GIT\TestBuffer\DynamicBuffer.js:59:45
    456    2.9%    2.9%  LazyCompile: *ToPositiveInteger native runtime.js:672:27
    375    2.3%    2.4%  Stub: GrowArrayElementsStub
    354    2.2%    2.2%  LazyCompile: *DynamicBuffer.setUint8 G:\GIT\TestBuffer\DynamicBuffer.js:34:45
    302    1.9%    1.9%  Stub: FastNewClosureStub
    294    1.8%    1.9%  LazyCompile: ~<anonymous> G:\GIT\TestBuffer\BinaryWriter.js:19:33
    283    1.8%    1.8%  LazyCompile: *BinaryWriter.writeUInt8 G:\GIT\TestBuffer\BinaryWriter.js:17:46
    275    1.7%    1.7%  Stub: FastNewContextStub
    253    1.6%    1.6%  LazyCompile: *BinaryWriter.toBuffer G:\GIT\TestBuffer\BinaryWriter.js:118:44
    218    1.4%    1.4%  LazyCompile: *<anonymous> G:\GIT\TestBuffer\app.js:1:11
    210    1.3%    1.3%  Stub: CallFunctionStub_Args1
    198    1.2%    1.3%  LazyCompile: ~<anonymous> G:\GIT\TestBuffer\BinaryWriter.js:59:33
    179    1.1%    1.1%  LazyCompile: *DynamicBuffer.setUint32 G:\GIT\TestBuffer\DynamicBuffer.js:54:46
    176    1.1%    1.1%  LazyCompile: *setInt32 native typedarray.js:1900:28
    171    1.1%    1.1%  LazyCompile: *DynamicBuffer.setUint16 G:\GIT\TestBuffer\DynamicBuffer.js:44:46
    162    1.0%    1.0%  LazyCompile: *BinaryWriter.writeInt32 G:\GIT\TestBuffer\BinaryWriter.js:57:46
```

As you can see, average execution time of DynamicBuffer code is much higher than BinaryWriter code.

Function | DynamicBuffer execution time | BinaryWriter execution time
--- | --- | ---
build/toBuffer | 5.1% | 1.6%
setInt32/writeInt32 | 2.9% | 1.0%
setUint8/writeUInt8 | 2.2% | 1.8%

**Comment by Barbosik** [comment]
hm, with these tests, I found that there is a performance bottle neck with memory allocation in BinaryWriter. It happens when there is more than 12703 cells in the packet (when packet is longer than 228654 bytes).

I think it's the reason why Luka967 has worse results for BinaryWriter. He using 100000 writes of uint32, its packet with 400000 bytes length (100000*4).

**Comment by Barbosik** [comment]
added fix for memory allocation and improved BinaryWriter performance for a little.
It's just fantastic! :) Server works with no lags with up to 30000-50000 cells on the client screen :)
Currently fix will be tested on test server. Because it uses unsafe approach to build packets, so there is need to test if all works as expected.

**Comment by Luka967** [comment]
Great to hear this helped :+1:

**Comment by NatsuTheGreat** [comment]
@Barbosik Can you share this fix?

**Comment by Barbosik** [comment]
I will commit it later. I replaced some node.js functions with custom implementation. Some kind of hack :)
So, there is need to make sure that it will not leads to bugs. It improves BinaryWriter performance for about 10%.
And the second fix is shared buffer for cell updates. It improves performance very significantly.
Especially for large cell update packets (> 12000 cells). Currently so large update will leads to lags due to slow memory allocation. With shared buffer it doesn't need to allocate memory at all, because it reuse already allocated memory.

**Comment by NatsuTheGreat** [comment]
Alright, can't wait for the commit :)

**Comment by Barbosik** [comment]
first change has some issues, so I reworked BinaryWriter for a little, here we are:

Each pass = 10 packets with 100000 cells in each packet:
```
[DEBUG] sanity1 passed
[DEBUG] sanity2 passed
[INFO ] ===PASS 01===
[INFO ] DynamicBuffer: 2453.146 [ms] (check=160)
[INFO ] BinaryWriter:  159.861 [ms] (check=160)
[INFO ] ===PASS 11===
[INFO ] DynamicBuffer: 2441.320 [ms] (check=160)
[INFO ] BinaryWriter:  135.806 [ms] (check=160)
[INFO ] ===PASS 21===
[INFO ] DynamicBuffer: 2423.339 [ms] (check=160)
[INFO ] BinaryWriter:  138.748 [ms] (check=160)
[DEBUG] process.exit(0)
```

**Comment by NatsuTheGreat** [comment]
Oh damn! That is fast!! I usually get 140-+ MS with 1000 bots, and overtime it gets really high. but 10 packets with 100000 cells in each of the packets, and getting around 140, is just crazy good!

**Comment by Barbosik** [comment]
I'm talking about this test code: https://github.com/Barbosik/MultiOgar/issues/203#issuecomment-233177499
It's not about server update time.

**Comment by NatsuTheGreat** [comment]
Well that's still really fast.

**Comment by Barbosik** [comment]
Here is test result for 3 packets with 1'000'000 cells in each packet (each packet length=18 megabytes):
```
[INFO ] ===PASS 01===
[INFO ] DynamicBuffer: 12389.442 [ms] (check=80)
[INFO ] BinaryWriter:  674.418 [ms] (check=80)
[INFO ] ===PASS 11===
[INFO ] DynamicBuffer: 12287.768 [ms] (check=80)
[INFO ] BinaryWriter:  602.102 [ms] (check=80)
[INFO ] ===PASS 21===
[INFO ] DynamicBuffer: 12266.858 [ms] (check=80)
[INFO ] BinaryWriter:  597.388 [ms] (check=80)
[DEBUG] process.exit(0)
```

**Comment by NatsuTheGreat** [comment]
BinaryWrite is practically twice as fast!

**Comment by Barbosik** [comment]
I'm talking about new BinaryWriter, I just implemented it. Currently I'm testing it on bubble-wars.tk:443
It already pass minion attack - worked very smooth even with 128 players on the server (server update time was 20 ms).

**Comment by NatsuTheGreat** [comment]
You are good!

**Comment by acydwarp** [comment]
Amazing speeds :)

You can probably push a bit more performance out of it by replacing `new Buffer()` with `Buffer.allocUnsafe()` in BinaryWriter and use `Buffer.from()` in BinaryReader. It requires Node.js 6 or higher. `new Buffer()` does too many things and is actually deprecated in Node.js.

**Comment by Barbosik** [comment]
yeah, throughput of new code is about 5'000'000 cells per second. It's for about 90 megabytes per second :)

**Comment by NatsuTheGreat** [comment]
@acydward Perfect. I have node v6.3.0, i'll use your suggestion.

@Barbosik Do you plan on sharing the new BinaryWriter?

**Comment by acydwarp** [comment]
Yeah I definitely recommend upgrading, there are many buffer performance enhancements in Node.js 6 and higher, including security patches and an upgraded v8 engine. You can view the changes in the Node.js [changelog](https://github.com/nodejs/node/blob/master/CHANGELOG.md).

**Comment by Luka967** [comment]
I made some experiments to DynamicBuffer today and made comparison perfomance tests. It's ~10% faster with creating large packets (more than around 200000 bytes) and ~10% slower with creating smaller ones. I'm astonished to see slower creation at small packets because all I did was make a shared byte calculator (the temp variables).

I also finally made a working quadtree and implemented it in a few places. :3
Currently making a large improvement on perfomance using setTimeout magic again. It looks like it has promising results but the effects are in almost all files.
